api/pkg/transformer/feast/transformer.go

```diff
@@ -337,62 +339,79 @@ func getExpressionExtractor(entity *transformer.Entity) string {
 	return entity.GetUdf()
 }
 
-func (t *Transformer) buildEntitiesRequest(ctx context.Context, request []byte, configEntities []*transformer.Entity) ([]feast.Row, error) {
+func (t *Transformer) buildEntitiesRequest(ctx context.Context, request []byte, configEntities []*transformer.Entity, tableName string) ([]feast.Row, error) {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "feast.buildEntitiesRequest")
+	span.SetTag("table.name", tableName)
 	defer span.Finish()
 
-	var entities []feast.Row
 	var nodesBody interface{}
 	err := json.Unmarshal(request, &nodesBody)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, configEntity := range configEntities {
-		switch configEntity.Extractor.(type) {
-		case *transformer.Entity_JsonPath:
-			_, ok := t.compiledJsonPath[configEntity.GetJsonPath()]
-			if !ok {
-				c, err := jsonpath.Compile(configEntity.GetJsonPath())
-				if err != nil {
-					return nil, fmt.Errorf("unable to compile jsonpath for entity %s: %s", configEntity.Name, configEntity.GetJsonPath())
-				}
-				t.compiledJsonPath[configEntity.GetJsonPath()] = c
-			}
-		}
+	var allSeries [][]*types.Value
+	maxLength := 1
 
+	for k, configEntity := range configEntities {
 		expressionExtractor := getExpressionExtractor(configEntity)
+
 		vals, err := getValuesFromJSONPayload(nodesBody, configEntity, t.compiledJsonPath[configEntity.GetJsonPath()], t.compiledUdf[expressionExtractor])
 		if err != nil {
 			return nil, fmt.Errorf("unable to extract entity %s: %v", configEntity.Name, err)
 		}
 
-		if len(entities) == 0 {
-			for _, val := range vals {
-				entities = append(entities, feast.Row{
-					configEntity.Name: val,
-				})
-			}
-		} else {
-			newEntities := []feast.Row{}
-			for _, entity := range entities {
-				for _, val := range vals {
-					newFeastRow := feast.Row{}
-					for k, v := range entity {
-						newFeastRow[k] = v
-					}
+		seriesLength := len(vals)
 
-					newFeastRow[configEntity.Name] = val
-					newEntities = append(newEntities, newFeastRow)
-				}
-			}
-			entities = newEntities
+		if seriesLength != 1 && maxLength != 1 && seriesLength != maxLength {
+			return nil, fmt.Errorf("entity %s has different dimension", configEntities[k].Name)
+		}
+
+		if seriesLength > maxLength {
+			maxLength = seriesLength
+		}
+
+		allSeries = append(allSeries, vals)
+	}
+
+	entities := make([]feast.Row, maxLength)
+	for s, series := range allSeries {
+		entityName := configEntities[s].Name
+
+		if len(series) == 1 {
+			entities = t.broadcastSeries(entityName, series, entities)
+		} else {
+			entities = t.addSeries(entityName, series, entities)
 		}
 	}
 
 	return entities, nil
 }
 
+func (t *Transformer) broadcastSeries(entityName string, series []*types.Value, entities []feast.Row) []feast.Row {
+	for e := range entities {
+		newFeastRow := feast.Row{}
+		for k, v := range entities[e] {
```

Comment:
I think this part will make the whole `buildEntitiesRequest` slower as it will be 3 nested for loop. In worst case it will be O(m*n^2), m: number of series, n: number of entities, it might not be good when there are a lot of entities.

Reply:
This can be simply
```
	for _, entity := range entities {
		entity[entityName] = series[0]
	}
	return entities
```